File: 003-ArrayDiff/array_diff.go
// array_diff.go

package main

func arrayDiff(a, b []int) []int {
	// creating a map to mark elemenst present in `b`
	bMap := make(map[int]bool)
	for _, num := range b {
		bMap[num] = true
	}

	// Use a map to store unique elements from `a` that are not in `b`
	uniqueResult := make(map[int]bool)
	result := [] int{}
	for _, num := range a {
		if !bMap[num] && !uniqueResult[num] {
			uniqueResult[num] = true
			result = append(result, num)
		}
	}
	return result
}

File: 003-ArrayDiff/arrayDiff.cpp
#include <vector>
#include <unordered_set>

std::vector<int> arrayDiff(const std::vector<int>& a, const std::vector<int>& b) {
    // creating set with elements of `b for quick looku
    std::unordered_set<int> bSet(b.begin(), b.end());

    // Use a cetor to store the result with unique elements, preserving order
    std::vector<int> result;
    std::unordered_set<int> seen;
    for (int num : a) {
        if(bSet.find(num) == bSet.end() && seen.find(num) == seen.end()) {
            result.push_back(num);
            seen.insert(num);
        }
    }
    return result;
}